src/analyze.rs

```diff
@@ -0,0 +1,80 @@
+
+use crate::{base::Res, listen::get_notes_from_audio_data, note::Note};
+
+/// Retrieve a list of notes which are guessed from the given audio clip.
+pub fn get_notes_from_audio_file(file: impl AsRef<Path>, start: Option<Duration>, end: Option<Duration>) -> Res<Vec<Note>> {
+    let path = file.as_ref();
+    let file = File::open(path)?;
+    let start = start.unwrap_or_default();
+    let decoder = Decoder::new(file)?;
+    let decoder = decoder.skip_duration(start).convert_samples();
+    let sample_rate = decoder.sample_rate() as usize;
+    let samples: Vec<_> = if let Some(end) = end { decoder.take_duration(end - start).collect() } else { decoder.collect() };
+    let clip_length = sample_rate.div_ceil(samples.len()).try_into()?;
+    get_notes_from_audio_data(&samples, clip_length)
+}
+
+/// Play the given segment of an audio file. Used to preview a clip before guessing notes from it.
+pub fn preview_audio_file_clip(file: impl AsRef<Path>, start: Option<Duration>, end: Option<Duration>) -> Res<()> {
+    let file = File::open(file)?;
+    preview_audio_clip(file, start, end)
+}
+
+/// Play the given segment of an audio stream. Used to preview a clip before guessing notes from it.
+pub fn preview_audio_clip(stream: impl Read + Seek + Send + Sync + 'static, start: Option<Duration>, end: Option<Duration>) -> Res<()> {
+    let start = start.unwrap_or_default();
+    let decoder = Decoder::new(stream)?;
+    let decoder = decoder.skip_duration(start).convert_samples();
+    let (_stream, stream_handle) = OutputStream::try_default()?;
+    if let Some(end) = end {
+        stream_handle.play_raw(decoder.take_duration(end - start))?;
+        sleep(end - start);
+    } else if let Some(duration) = decoder.total_duration() {
+        stream_handle.play_raw(decoder)?;
+        sleep(duration);
+    } else {
+        let channels = decoder.channels();
+        let sample_rate = decoder.sample_rate() as f32;
+        let samples: Vec<_> = decoder.collect();
+        let time = Duration::from_secs((samples.len() as f32 / sample_rate).ceil() as u64);
+        stream_handle.play_raw(SamplesBuffer::new(channels, sample_rate as u32, samples))?;
+        sleep(time);
+    }
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use std::path::PathBuf;
+
+    use crate::note::{BFive, CFive, DSix, EFlatSeven, GEight};
+
+    use super::*;
+
+    #[test]
+    fn test_preview_audio_clip() {
+        preview_audio_file_clip(PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests/MSC_AnaSaurus[shrt]-Cmaj7-9.wav"), None, None).unwrap();
+    }
+
+    #[test]
+    fn test_get_notes_from_audio_file() {
+        let path = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests/MSC_AnaSaurus[shrt]-Cmaj7-9.wav");
+        let notes = get_notes_from_audio_file(path, None, None).expect("notes");
+        println!("{notes:?}");
+        assert!(notes.contains(&CFive));
+        assert!(notes.contains(&EFlatSeven));
+        assert!(notes.contains(&GEight));
```

Comment:
Can we get an audio file that is a little more straightforward?  Like something that is a regular Cm7b9, or something like that?

Mostly concerned about that G8 coming though.  Do you know why that G8 is there?

Reply:
Sorry, I just grabbed a sample from some pack on the internet. Shouldn't probably do that anyway for copyright reasons. I have no idea why there is a G8, or, really, if it's even there. I thought that the issue with misdetection earlier was because I generated those sample files from a DAW that was running in WINE. I downloaded LMMS to try to generate a fresh file from a synth that was not going through wine, and now the tests fail again:

```
failures:

---- analyze::tests::test_get_notes_from_audio_file stdout ----
[Note { octave: Six, named_pitch: DFlat }, Note { octave: Five, named_pitch: A }, Note { octave: Five, named_pitch: GFlat }]
thread 'analyze::tests::test_get_notes_from_audio_file' panicked at 'assertion failed: notes.contains(&AFour)', src/analyze.rs:72:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- analyze::tests::test_get_notes_from_mp3_file stdout ----
[Note { octave: Five, named_pitch: G }, Note { octave: Six, named_pitch: D }, Note { octave: Five, named_pitch: BFlat }]
thread 'analyze::tests::test_get_notes_from_mp3_file' panicked at 'assertion failed: notes.contains(&AFour)', src/analyze.rs:83:9
```

Here is the [LMMS project](https://github.com/twitchax/kord/files/10536503/lmms-project.zip) in a zip file, perhaps you could do some testing to figure out why the analyzed notes aren't right? Does the test file *sound like* an Am chord compared to some instrument you have in tune?

I've also pushed my changes to this branch, including the tests that aren't working and the example files.

Reply:
Ok, let's keep what you have for now, and I will look into getting a better sample.

Can you address the other comments?